libbeat/autodiscover/providers/kubernetes/kubernetes.go

```diff
@@ -88,53 +86,56 @@ func (p *Provider) Start() {
 }
 
 func (p *Provider) emit(pod *kubernetes.Pod, flag string) {
+	// Emit events for all containers
+	p.emitEvents(pod, flag, pod.Spec.Containers, pod.Status.ContainerStatuses)
+
+	// Emit events for all initContainers
+	p.emitEvents(pod, flag, pod.Spec.InitContainers, pod.Status.InitContainerStatuses)
+}
+
+func (p *Provider) emitEvents(pod *kubernetes.Pod, flag string, containers []kubernetes.Container,
+	containerstatuses []kubernetes.PodContainerStatus) {
 	host := pod.Status.PodIP
-	containerIDs := map[string]string{}
 
-	// Emit pod container IDs
-	for _, c := range append(pod.Status.ContainerStatuses, pod.Status.InitContainerStatuses...) {
-		cid := c.GetContainerID()
+	// Collect all container IDs and runtimes from status information. Kubernetes has both docker and rkt
+	containerIDs := map[string]string{}
+	runtimes := map[string]string{}
+	for _, c := range containerstatuses {
+		cid, runtime := c.GetContainerIDWithRuntime()
 		containerIDs[c.Name] = cid
+		runtimes[c.Name] = runtime
+	}
 
+	// Emit container and port information
+	for _, c := range containers {
 		cmeta := common.MapStr{
-			"id":    cid,
-			"name":  c.Name,
-			"image": c.Image,
+			"id":      containerIDs[c.Name],
+			"name":    c.Name,
+			"image":   c.Image,
+			"runtime": runtimes[c.Name],
 		}
-
-		// Metadata appended to each event
 		meta := p.metagen.ContainerMetadata(pod, c.Name)
 
 		// Information that can be used in discovering a workload
 		kubemeta := meta.Clone()
 		kubemeta["container"] = cmeta
 
-		// Emit container info
-		p.publish(bus.Event{
-			flag:         true,
-			"host":       host,
-			"kubernetes": kubemeta,
-			"meta": common.MapStr{
-				"kubernetes": meta,
-			},
-		})
-	}
+		// Pass annotations to all events so that it can be used in templating and by annotation builders.
+		kubemeta["annotations"] = pod.GetMetadata().Annotations
 
-	// Emit pod ports
-	for _, c := range pod.Spec.Containers {
-		cmeta := common.MapStr{
-			"id":    containerIDs[c.Name],
-			"name":  c.Name,
-			"image": c.Image,
+		// Without this check there would be overlapping configurations with and without ports.
```

Comment:
I'm ok with this change but could you open it in a separate PR? this one could grow a lot

Reply:
i have removed this. i will add it in once this PR is merged.